### PR TITLE
Add trainer integration test for llava to ensure accelerate autocasting works correctly

### PR DESCRIPTION
**Pull request by frasermince**
# What does this PR do?
This PR adds a new integration test to ensure the accelerate autocasting is working correctly. This came out of a discussion found [here](https://github.com/huggingface/transformers/pull/29721#discussion_r1567192103) and that PR should probably be merged first (or this one merged into that one).

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
      Not an issue but another PR that should probably be merged first: https://github.com/huggingface/transformers/pull/29721#discussion_r1567192103
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/models/llava/test_trainer_llava.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import unittest
```
Comment (frasermince) [inline tests/models/llava/test_trainer_llava.py:1]: This file should be the only new change. That will be clearer once the other is merged. Usually I would set it up to merge this one into the first but this feels a bit difference since the changes are on a fork.

File: tests/models/llava/test_trainer_llava.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+            "USER: <image>\nWhat is this?\nASSISTANT:",
+        ]
+        image1 = Image.open(requests.get("https://llava-vl.github.io/static/images/view.jpg", stream=True).raw)
+        image2 = Image.open(requests.get("http://images.cocodataset.org/val2017/000000039769.jpg", stream=True).raw)
+
+        inputs = self.processor(prompts, images=[image1, image2], return_tensors="pt", padding=True)
+
+        output = model(**inputs)
+        expected_slice = torch.tensor(
+            [[-3.5664, -3.5625, -0.4309], [-5.8242, -5.6914, -1.3242], [-5.4805, -5.9375, 1.1465]],
```
Comment (frasermince) [inline tests/models/llava/test_trainer_llava.py:111]: I saw this pattern in the llava next tests. These came from the use of this model before training. Not quite sure if this is correct so please let me know if there is something else we want to test. Perhaps the trained model before applying the downcasting change?

File: tests/models/llava/test_trainer_llava.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        train_dataset = Dataset.from_generator(image_prompt_generator).map(process_image_prompt)
+        bits = 4
+        bits_and_bytes_config = BitsAndBytesConfig(
+            load_in_4bit=True,
+        )
+        model = LlavaForConditionalGeneration.from_pretrained(
+            "llava-hf/bakLlava-v1-hf", quantization_config=bits_and_bytes_config
+        )
+        adapter_name = "lora_default"
+        peft_config = LoraConfig(
```
Comment (frasermince) [inline tests/models/llava/test_trainer_llava.py:79]: I am somewhat unclear on where we test the slow tests but I assume there is some limit on memory so I tried to give a reasonable LORA for this test. If you think there is a simpler or more idiomatic way to do this test please let me know.

File: tests/models/llava/test_trainer_llava.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        )
+
+    def tearDown(self):
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    @slow
+    @require_bitsandbytes
+    def test_model_trainer_integration_test(self):
+        def image_prompt_generator():
```
Comment (frasermince) [inline tests/models/llava/test_trainer_llava.py:44]: Not entirely sure this is the simplest or most idiomatic way to create this test dataset so please let me know if there is a better way.

File: tests/models/llava/test_trainer_llava.py (hunk 2 above)
Comment (frasermince) [reply tests/models/llava/test_trainer_llava.py:107]: Also note these do not yet pass `allclose`. I wanted to go ahead and open this PR to generate discussion around what the right thing to test is.

**Comment by amyeroberts** [comment]
cc @muellerzr for first review

**Comment by frasermince** [comment]
Updated this now that the previous PR is merged! I am very concerned about OOMs being an issue here however. I think there's some open questions around:
1. How we ensure models are compatible with the trainer and accelerate
2. How we test training a model in CI given how memory intensive this can be

**Comment by amyeroberts** [comment]
Gentle ping @muellerzr, or possibly @SunMarc?

**Comment by amyeroberts** [comment]
Another ping @muellerzr @SunMarc

File: tests/models/llava/test_trainer_llava.py (hunk 3 above)
Comment (muellerzr) [reply tests/models/llava/test_trainer_llava.py:80]: I think this is fine for what we're doing!

**Review by muellerzr — APPROVED** [review]
Overall this is a very good test. Since autocasting is done "automagically" via accelerate, this tests it exactly how you should!

File: tests/models/llava/test_trainer_llava.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
+import gc
```
Comment (amyeroberts) [inline tests/models/llava/test_trainer_llava.py:1]: missing copyright header

File: tests/models/llava/test_trainer_llava.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+class LlavaForConditionalGenerationIntegrationTest(TestCasePlus):
+    def setUp(self):
+        super().setUp()
+        self.processor = AutoProcessor.from_pretrained(
+            "llava-hf/bakLlava-v1-hf", padding_side="left", truncation_side="right"
+        )
+
+    def tearDown(self):
+        gc.collect()
+        torch.cuda.empty_cache()
```
Comment (amyeroberts) [inline tests/models/llava/test_trainer_llava.py:41]: My (somewhat sparse) knowledge of empty_cache is that it's not meant to be used manually and can cause unintended / surprising behaviour: https://github.com/huggingface/transformers/issues/31372#issuecomment-2170354817

**Review by amyeroberts — APPROVED** [review]
Looks great - thanks for adding this! 

Once the conflicts are resolved I think we're good to go 

cc @zucchini-nlp for reference

File: tests/models/llava/test_trainer_llava.py (hunk 6 above)
Comment (muellerzr) [reply tests/models/llava/test_trainer_llava.py:41]: While normally I’d agree, if it’s in the tests it should be fine. That was in reference to it being in the actual Trainer code

**Review by zucchini-nlp — COMMENTED** [review]
Interesting finding, thanks for fixing and adding tests!

Out of curiosity, iiuc training llava and llava-next with HF Trainer and fp16/bf16 flags failed with dtype errors, before the fix was done. I am wondering about other llava-based models, this should mean that vipllava, llava-next-video and video-llava should fail with the same error because all follow similar architecture. But for llava-next-video I had a script with fp16 running w/o errors, would love to know your opinion on this

File: tests/models/llava/test_trainer_llava.py (hunk 2 above)
Comment (frasermince) [reply tests/models/llava/test_trainer_llava.py:112]: @muellerzr How are these allclose values found? Is it literally what the model is outputting now to determine if logits are changing in the future? Is it based on some original implementation?
